### PR TITLE
Remove `spack.target` from code

### PR DESCRIPTION
**Pull request by alalazo**
The `spack.target.Target` class is a weird entity, that is just needed to:

1. Sort microarchitectures in lists deterministically
2. Being able to use microarchitectures in hashed containers

This PR removes it, and uses `archspec.cpu.Microarchitecture` directly. To sort lists, we use a proper `key=` when needed. Being able to use `Microarchitecture` objects in sets is achieved by updating the external `archspec`.

We might clean the code further, in following PRs, by introducing microarchitecture ranges in `ArchSpec`, and simplifying the constrain etc. logic.

### CODE REVIEW

File: var/spack/repos/builtin/packages/neuron/package.py [hunk 1]
```diff
@@ -149,7 +150,7 @@ def cmake_args(self):
 
         # add cpu arch specific optimisation flags to CMake so that they are passed
         # to embedded Makefile that neuron has for compiling MOD files
-        compilation_flags = self.spec.architecture.target.optimization_flags(self.spec.compiler)
+        compilation_flags = optimization_flags(self.spec.compiler, self.spec.target)
         args.append(self.define("CMAKE_CXX_FLAGS", compilation_flags))
 
         return args
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neuron/package.py:156]: Just checked the neuron package, and nothing important - LGTM.

**Review by rbberger — APPROVED** [review]
LGTM for the lammps package.

**Comment by haampie** [comment]
```
lib/spack/spack/platforms/_platform.py:docstring of spack.platforms._platform.Platform.add_target:1: WARNING: py:class reference target not found: archspec.cpu.microarchitecture.Microarchitecture
```

**Review by matz-e — APPROVED** [review]

**Comment by alalazo** [comment]
The only failure seems unrelated to the PR: https://gitlab.spack.io/spack/spack/-/jobs/12614028

**Review by haampie — APPROVED** [review]

**Comment by haampie** [comment]
Adding `v0.22.3` because just archspec needs to be backported. Should've been a separate PR.

**Comment by alalazo** [comment]
The only conflict is the sha, that can also be solved in the backport PR. We can keep archspec updates always separate in the future. For what is worth, the changes in `archspec` here are because of this PR.

**Pull request merged**